ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -0,0 +1,250 @@
+   * @param messageBuf the bytes to be uploaded.
+   * @param size the number of bytes to upload.
+   * @throws Exception
+   */
+  private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size) throws Exception {
+    boolean performUpload = messageInfo.getExpirationTimeInMs() == Utils.Infinite_Time && !messageInfo.isDeleted();
+    // May need to encrypt buffer before upload
+    if (performUpload) {
+      BlobId blobId = (BlobId) messageInfo.getStoreKey();
+      cloudDestination.uploadBlob(blobId, size, new ByteBufferInputStream(messageBuf));
+    }
+  }
+
+  @Override
+  public void delete(MessageWriteSet messageSetToDelete) throws StoreException {
+    checkStarted();
+    checkDuplicates(messageSetToDelete);
+
+    try {
+      for (MessageInfo msgInfo : messageSetToDelete.getMessageSetInfo()) {
+        BlobId blobId = (BlobId) msgInfo.getStoreKey();
+        cloudDestination.deleteBlob(blobId);
+      }
+    } catch (Exception ex) {
+      throw new StoreException(ex, StoreErrorCodes.IOError);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * Currently, the only supported operation is to set the TTL to infinite (i.e. no arbitrary increase or decrease)
+   * @param messageSetToUpdate The list of messages that need to be updated
+   * @throws StoreException if there is a problem persisting the operation in the store.
+   */
+  @Override
+  public void updateTtl(MessageWriteSet messageSetToUpdate) throws StoreException {
+    checkStarted();
+    // This is currently a no-op because we skipped uploading the blob on receipt of the PUT record since it had a TTL.
+  }
+
+  @Override
+  public FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws StoreException {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public Set<StoreKey> findMissingKeys(List<StoreKey> keys) throws StoreException {
+    checkStarted();
+    // Check existence of keys in cloud container
+    Set<StoreKey> missingKeys = new HashSet<>();
+    try {
+      for (StoreKey key : keys) {
+        BlobId blobId = (BlobId) key;
+        if (!cloudDestination.doesBlobExist(blobId)) {
+          missingKeys.add(key);
+        }
+      }
+    } catch (Exception ex) {
+      throw new StoreException(ex, StoreErrorCodes.IOError);
+    }
+    return missingKeys;
+  }
+
+  @Override
+  public StoreStats getStoreStats() {
+    return null;
+  }
+
+  @Override
+  public boolean isKeyDeleted(StoreKey key) throws StoreException {
+    checkStarted();
+    // Return false for now, because we don't track recently deleted keys
+    // This way, the replica thread will replay the delete resulting in a no-op
+    // TODO: Consider LRU cache of recently deleted keys
+    return false;
+  }
+
+  @Override
+  public long getSizeInBytes() {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return false;
```

Comment:
this should be possible to implement?

Reply:
Yes. I need to add some state so the store can ask the destination for its current state when it starts up.  I'll add a TODO.